smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/StructuredMemberWriter.java

```diff
@@ -65,6 +70,140 @@ void writeMembers(TypeScriptWriter writer, Shape shape) {
+            // writer.write("item => item");
+        }
+    }
+
+    /**
+     * Recursively writes filterSensitiveLog for MapShape.
+     */
+    void writeMapFilterSensitiveLog(TypeScriptWriter writer, MemberShape mapMember) {
+        // Reducer is common to all shapes.
+        writer.openBlock("(acc: any, [key, value]: [string, $T]) => ({", "}), {}",
+            symbolProvider.toSymbol(mapMember),
+            () -> {
+                writer.write("...acc,");
+                Shape memberShape = model.expectShape(mapMember.getTarget());
+                if (memberShape instanceof StructureShape) {
+                    // Call filterSensitiveLog on Structure.
+                    writer.write("[key]: $T.filterSensitiveLog(value),",
+                        symbolProvider.toSymbol(mapMember));
+                } else if (memberShape instanceof CollectionShape) {
+                    writer.openBlock("[key]: value.map(", "),",
+                        () -> {
+                            MemberShape collectionMember = ((CollectionShape) memberShape).getMember();
+                            writeCollectionFilterSensitiveLog(writer, collectionMember);
+                        }
+                    );
+                } else if (memberShape instanceof MapShape) {
+                    writer.openBlock("[key]: Object.entries(value).reduce(", "),",
+                        () -> {
+                            MemberShape nestedMapMember = ((MapShape) memberShape).getValue();
+                            writeMapFilterSensitiveLog(writer, nestedMapMember);
+                        }
+                    );
+                } else {
+                    // This path should not reach because of recursive isIterationRequired.
+                    throw new CodegenException(String.format(
+                        "MapFilterSensitiveLog attempted for %s while it was not required",
+                        memberShape.getType()
+                    ));
+                    // For quick-fix in case of high severity issue:
+                    // comment out the exception above and uncomment the line below.
+                    // writer.write("[key]: value,");
+                }
+            }
+        );
+    }
+
+    void writeFilterSensitiveLog(TypeScriptWriter writer, Shape shape) {
+        writer.write("...obj,");
+        for (MemberShape member : members) {
+            Shape memberShape = model.expectShape(member.getTarget());
+            String memberName = TypeScriptUtils.sanitizePropertyName(symbolProvider.toMemberName(member));
+            if (member.getMemberTrait(model, SensitiveTrait.class).isPresent()) {
+                // member is Sensitive, hide the value.
+                writer.write("...(obj.$1L && { $1L: SENSITIVE_STRING }),", memberName);
+            } else if (memberShape instanceof StructureShape) {
+                // Call filterSensitiveLog on Structure.
```

Comment:
You probably want to also check `isIterationRequired` here in case the explicit `return true` changes for any condition in the future.

Reply:
For a structure at top level, `filterSensitiveLog` should always be called ([TypeScript playground](https://www.typescriptlang.org/play/?ssl=20&ssc=37&pln=20&pc=53#code/MYewdgzgLgBAygUQHJwJIBVUDUEH07oBKqSA4jALwwBEAVPXAKaQCWULAbo6mAGYgAnALYBDduHq1qAbgBQslmCiMBvEcEYwAqhBUwA3rJgwArroFgRQxgC4Y0AYoDmc4wAcRECAHdBAEzsHZzkAX3lLawgPDW1zAyMYRgAPN0FYUEhYXhYAG2UBJlZ2LgAZECdKGAAKEAAjACs7HRUASkoAPmrDY2MAOn66+oAaBPdPH387RBQMbDwCYjIEkJbZMNkM6BhlaGaBJriqbtNzCNsaMxUzgEZqEbGvXwEAmg9H-1uRkLkN8AgQHKMXo5cpVPa9bJ5FSFCBsTiMMpOKo7KB7FotORAA))

The `isIterationRequired` is only required for CollectionShape and MapShape at top level.

Reply:
Added explicit check in refactor commit https://github.com/awslabs/smithy-typescript/pull/170/commits/f8ac27097b191475420c1a441f94074b13c48b86